source/s3_client.c

```diff
@@ -0,0 +1,168 @@
+    (void)self;
+
+    struct aws_allocator *allocator = aws_py_get_allocator();
+
+    PyObject *bootstrap_py;
+    PyObject *credential_provider_py;
+    PyObject *tls_options_py;
+    PyObject *on_shutdown_py;
+    const char *region;
+    Py_ssize_t region_len;
+    uint64_t part_size;
+    double throughput_target_gbps;
+    if (!PyArg_ParseTuple(
+            args,
+            "OOOOs#Kd",
+            &bootstrap_py,
+            &credential_provider_py,
+            &tls_options_py,
+            &on_shutdown_py,
+            &region,
+            &region_len,
+            &part_size,
+            &throughput_target_gbps)) {
+        return NULL;
+    }
+
+    struct aws_client_bootstrap *bootstrap = aws_py_get_client_bootstrap(bootstrap_py);
+    if (!bootstrap) {
+        return NULL;
+    }
+
+    struct aws_credentials_provider *credential_provider = NULL;
+    if (credential_provider_py != Py_None) {
+        credential_provider = aws_py_get_credentials_provider(credential_provider_py);
+    }
+
+    struct aws_signing_config_aws signing_config;
+    AWS_ZERO_STRUCT(signing_config);
+
+    struct aws_byte_cursor region_cursor = aws_byte_cursor_from_array((const uint8_t *)region, region_len);
+
+    if (credential_provider) {
+        aws_s3_init_default_signing_config(&signing_config, region_cursor, credential_provider);
+    }
+
+    struct s3_client_binding *s3_clinet = aws_mem_calloc(allocator, 1, sizeof(struct s3_client_binding));
+    if (!s3_clinet) {
+        return PyErr_AwsLastError();
+    }
+
+    /* From hereon, we need to clean up if errors occur */
+    struct aws_tls_connection_options *tls_options = NULL;
+    if (tls_options_py != Py_None) {
+        tls_options = aws_py_get_tls_connection_options(tls_options_py);
+        if (!tls_options) {
+            goto client_init_failed;
+        }
+    }
+
+    struct aws_s3_client_config s3_config = {
+        .region = aws_byte_cursor_from_array((const uint8_t *)region, region_len),
+        .client_bootstrap = bootstrap,
+        .signing_config = credential_provider ? &signing_config : NULL,
+        .part_size = part_size,
+        .tls_connection_options = tls_options,
+        .throughput_target_gbps = throughput_target_gbps,
+        .shutdown_callback = s_s3_client_shutdown,
+        .shutdown_callback_user_data = s3_clinet,
+    };
+
+    s3_clinet->native = aws_s3_client_new(allocator, &s3_config);
+    if (s3_clinet->native == NULL) {
+        PyErr_SetAwsLastError();
+        goto client_init_failed;
+    }
+
+    PyObject *capsule = PyCapsule_New(s3_clinet, s_capsule_name_s3_client, s_s3_client_capsule_destructor);
+    if (!capsule) {
+        goto capsule_new_failed;
+    }
+
+    s3_clinet->on_shutdown = on_shutdown_py;
+    Py_INCREF(s3_clinet->on_shutdown);
+
+    return capsule;
+
+capsule_new_failed:
+    aws_s3_client_release(s3_clinet->native);
+client_init_failed:
+    aws_mem_release(allocator, s3_clinet);
+    return NULL;
```

Comment:
Ugh. setup & shutdown error-handling code is always so fragile because you need to account for setup failing halfway through. For these reasons, I've come to prefer creating the capsule as soon as possible, rather than creating it as the last step. That way, if anything goes wrong I can just decref the capsule and let its destructor clean things up, even if setup only partially succeeded. Personally, I've found it simpler to write shutdown code that can handle a partially set up thing, instead of setup code that can shut down a partially setup thing, in addition to regular shutdown code.

So the order of fallible setup operations would be like:
```
s3_client = aws_mem_calloc()
capsule = PyCapsule_New(...)
if (!capsule)
    aws_mem_destroy(s3_client)
    return NULL

s3_client->on_shutdown = on_shutdown_py
Py_INCREF(on_shutdown)

s3_client->native = aws_s3_client_new()
if (!native)
    PyErr_SetAwsLastError()
    goto error;
return capsule
error:
Py_DECREF(capsule);
return NULL
```

but cleanup needs to check for the case that the native S3 client never actually existed
```
--- s_capsule_destructor() ---
aws_s3_client_release(client->native) // yeah, may be NULL, but release() can handle it
client->release_called=true;
s_destroy_if_ready(client)

--- s_on_shutdown() ---
fire callbacks
Py_CLEAR(client->on_shutdown)
client->shutdown_called = true
s_destroy_if_ready(client)

--- s_destroy_if_ready() ---
if (client->native && (!client->shutdown_called || !client->release_called))
    // native client successfully created, but not ready to clean up yet
    return

Py_XDECREF(client->on_shutdown) // just in case native never existed and shutdown never happened
aws_mem_release(client)
```